### ISSUE TITLE
Generator return type always widened

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#57910]
### 🔎 Search Terms

generator literal widen

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about (I am the faq author)

### ⏯ Playground Link

https://www.typescriptlang.org/play?target=99&jsx=0#code/CYUwxgNghgTiAEYD2A7AzgF3gQQFzwCJsCBuAKAHoL4AlARnwHEQUQYoMkYAeTGASxQBzADTwAbkn7AxAVxSgAZoJDAAfJWoBRAB4AHcBlVMWbDl25ECYydLkKQy1urLJ0WevAC88ABSL5MAx+VAAqXwBKeABvMnh4+ABPfhAIYFCccgBfCMjyKloAJhNWdk4ePkFRCSkZeHklFRcC3QMg43hmUvMeKxta+0bnDTdMIu8-AJQgkJRwqNiEpJS0jKt4KDREVExs3Ij87RgYLk14VsMOgDkkeDYTmFcdjwBmErNyy2J+u3qHJ1Uagm-kCwTCkRicQSyVS6UIxA2W1GGD2eTOWmOpxa+kuwHwNzumMeyNoABZ3mULH0ar8Go4mkCfCDpmC5hDFtCVnDsKiDkA

### 💻 Code

```ts
declare const A: "A";
// R1: Generator<string, void, undefined>
// Expected: Generator<"A", void, undefined>
const R1 = (function*() {
    yield* A;
})();
// R2: Generator<string, void, undefined>
// Expected: Generator<"A", void, undefined>
const R2 = (function*() {
    yield* "A" as const;
})();
// Error
// Expected: No error
const R3: Generator<"A", void, undefined> = (function*() {
    yield* "A" as const;
})();
// Error
// Expected: No error
const R4: Generator<"A", void, undefined> = (function*() {
    yield* A;
})();
```

### 🙁 Actual behavior

In all cases, the literal return is widened. This also extends to places like
```ts
// R5: Generator<string, not Generator<"A" | "B"
const R5 = (function*() {
    yield* "A" as const;
    yield* "B" as const;
})();
```
where our normal "union of literals" logic should have happened

### 🙂 Expected behavior

Follow the same widening behavior as normal function returns

### Additional information about the issue

Discussed at #23537, but with a contradictory outcome

**Comment by RyanCavanaugh** [comment]
Iterating "A" produces strings 🤦

**Comment by fatcerberus** [comment]
> Iterating "A" produces strings 🤦

`string` is such a weird animal in this regard.  It's the only type (AFAIK) that's idempotent under iteration.  A naive algorithm for recursive iteration could potentially loop forever.
